### PR TITLE
Platform/jessica/12289 support complicated boolean expressions

### PR DESCRIPTION
**Pull request by JessicaWNava**
This PR adds support for complicated boolean expressions 

Test Steps:
1. Make sure the tests still pass

## Changes
- Created a new Condition
- Integrated that condition into the existing framework, but had it circumvent the logic to get it to work. 

### Testing
- [ X] Tested locally?
- [ ] Ran `./prime test` or `./gradlew testSmoke` against local Docker ReportStream container?
- [ ] (For Changes to /frontend-react/...) Ran `npm run lint:write`? 
- [ ] Added tests?

### CODE REVIEW

**Review by mkalish — COMMENTED** [review]
This is a neat idea!  I think we'll need to tweak it a bit to work with JDK17 though, so let me know if you want to pair up on this.

File: src/main/java/io/github/linuxforhealth/core/expression/condition/CompoundAndOrCondition.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+                conditionStatementContainer = makeRecursive(conditionStatementContainer, contextVariables);
+            } else {
+                conditionStatementContainer = eval(conditionStatementContainer, conditionStatementContainer, contextVariables);
+            }
+        }
+
+        return conditionStatementContainer;
+    }
+
+    private String fixParens(String conditionStatement) {
```
Comment (mkalish) [inline src/main/java/io/github/linuxforhealth/core/expression/condition/CompoundAndOrCondition.java:52]: I'm having a bit of a hard time following what's happening with this function, could we add some documentation? It sort of seems like this is trying to deterministically handle a situation where there are still `&&` and `||` but no more parens and the code will favor  grouping the or?

Also looking at the coverage report it looks like only the `else` branch is executed

File: src/main/java/io/github/linuxforhealth/core/expression/condition/CheckNotNull.java [hunk 2]
```diff
@@ -31,7 +31,7 @@ public boolean test(Map<String, EvaluationResult> contextVariables) {
     EvaluationResult variable1 = ContextValueUtils.getVariableValuesFromVariableContextMap(var1,
         contextVariables, false, VariableUtils.isFuzzyMatch(var1));
     
-    return variable1 != null && !variable1.isEmpty();
+    return variable1 != null && !variable1.isEmpty() && !variable1.getValue().equals("null");
```
Comment (mkalish) [inline src/main/java/io/github/linuxforhealth/core/expression/condition/CheckNotNull.java:34]: Is this change strictly necessary?  Couldn't an HL7 value be literally the string `"null"` and then it should pass this check?

File: src/main/java/io/github/linuxforhealth/core/expression/condition/SimpleBooleanCondition.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
+package io.github.linuxforhealth.core.expression.condition;
+
+import io.github.linuxforhealth.api.Condition;
+import io.github.linuxforhealth.api.EvaluationResult;
+
+import java.util.Map;
+
+public class SimpleBooleanCondition implements Condition {
```
Comment (mkalish) [inline src/main/java/io/github/linuxforhealth/core/expression/condition/SimpleBooleanCondition.java:8]: Should we add some tests for this/what is this trying to handle?

**Review by mkalish — COMMENTED** [review]
Awesome stuff!  Neat idea on how to handle the problem.  Left a few comments, but main thing to do before merging would be try this in reportstream.  I hit an issue, but I think you just need to rebase your branch against the up to date version.

File: src/main/java/io/github/linuxforhealth/core/expression/condition/CompoundAndOrCondition.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+            conditionStatement = "(" + conditionStatement.substring(0, orIndex-2) + ") " + conditionStatement.substring(orIndex, conditionStatement.length());
+        } else {
+            conditionStatement = conditionStatement.substring(0, orIndex+2) + " (" + conditionStatement.substring(orIndex+2, conditionStatement.length()) + ")";
+        }
+        return conditionStatement;
+    }
+
+    private String eval(String conditionStatement, String group, Map<String, EvaluationResult> contextVariables) {
+        boolean goodToGo = false;
+        if(group.contains("||")) {
```
[the remark this answers is not in the document]
Comment (JessicaWNava) [reply src/main/java/io/github/linuxforhealth/core/expression/condition/CompoundAndOrCondition.java:65]: How so? They use different classes.

File: src/main/java/io/github/linuxforhealth/core/expression/condition/CheckNotNull.java (hunk 2 above)
Comment (JessicaWNava) [reply src/main/java/io/github/linuxforhealth/core/expression/condition/CheckNotNull.java:34]: Yes, it is necessary. The code will not work without it. I don't see why we wouldn't want to handle a string "null" as null anyways. It is null.

File: src/main/java/io/github/linuxforhealth/core/expression/condition/CompoundAndOrCondition.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (mkalish) [reply src/main/java/io/github/linuxforhealth/core/expression/condition/CompoundAndOrCondition.java:65]: Right, but don't they share a common interface that has `test` on it?

File: src/main/java/io/github/linuxforhealth/core/expression/condition/CheckNotNull.java (hunk 2 above)
Comment (mkalish) [reply src/main/java/io/github/linuxforhealth/core/expression/condition/CheckNotNull.java:34]: Right... but in Java/Kotlin 

```java
"null" == null // false
```

not necessarily advocating to remove this, but just noting that this change removes the ability to differentiate between the two

Comment (JessicaWNava) [reply src/main/java/io/github/linuxforhealth/core/expression/condition/CheckNotNull.java:34]: Yes, I understand that. I just don't see a case where we would not want to handle "null" as null.

Comment (mkalish) [reply src/main/java/io/github/linuxforhealth/core/expression/condition/CheckNotNull.java:34]: Yea, I think it's fine, but just thinking we should highlight this in the docs on conditions

File: src/main/java/io/github/linuxforhealth/core/expression/condition/CompoundAndOrCondition.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (JessicaWNava) [reply src/main/java/io/github/linuxforhealth/core/expression/condition/CompoundAndOrCondition.java:94]: Yes. I have updated the code, but I prefer the first way I did it since this introduces a bunch of potential exceptions.

File: src/main/java/io/github/linuxforhealth/core/expression/condition/CheckNotNull.java (hunk 2 above)
Comment (JessicaWNava) [reply src/main/java/io/github/linuxforhealth/core/expression/condition/CheckNotNull.java:34]: Where are the docs you are referring to located?

Comment (mkalish) [reply src/main/java/io/github/linuxforhealth/core/expression/condition/CheckNotNull.java:34]: I think its `TEMPLATING.md`

**Review by mkalish — APPROVED** [review]
LGTM!  Just the one documentation update before merging

File: src/main/java/io/github/linuxforhealth/core/expression/condition/CheckNotNull.java (hunk 2 above)
Comment (jalbinson) [reply src/main/java/io/github/linuxforhealth/core/expression/condition/CheckNotNull.java:34]: What if someone's last name is Null? Or they live on Null Street? Or theres some sort of reference to a null hypothesis (Just making up unlikely scenarios...)

It's not likely but will be a royal pain to debug if we end up running into an issue.

**Review by jalbinson — APPROVED** [review]
Some very cool stuff in here!

File: src/main/java/io/github/linuxforhealth/core/expression/condition/CheckNotNull.java (hunk 2 above)
Comment (JessicaWNava) [reply src/main/java/io/github/linuxforhealth/core/expression/condition/CheckNotNull.java:34]: This doesn't process the actual message content. This processes the condition statement that we as developers may put on a field to determine whether or not that field should be sent through in the message or not.
